### ISSUE TITLE
Perplexity evaluation too high for 1bitLLM/bitnet_b1_58-3B

### ISSUE THREAD

**Issue by MekkCyber** [microsoft/BitBLAS#47]
Hello Everyone, 

I am trying to evaluate the perplexity of 1bitLLM/bitnet_b1_58-3B, using the script available in integration/BitNet. However, I am getting a very high loss, and perplexity. Is it normal ?
```
avg_loss = 14.603133460411653: 100%|███████████████████████████████████████████████████████████████████████████████████████████████| 174/174 [02:04<00:00,  1.40it/s]
wikitext2 PPL: 24887.495944644015
[23059.264828634798, 24887.495944644015]
Avg PPL: 23973.38038663941
```

**Comment by LeiWang1999** [comment]
Thanks for your report @MekkCyber ! it's unusual, let me check it out and get back to you, thanks.

**Comment by igrekun** [comment]
Recent update might've broken something. Here are the results between simulated and bitblas matmul

```
t_input = torch.randn(1, 64).to("cuda")
t_linear = utils_quant.BitLinear(64, 128, bias=False).to("cuda")
with torch.no_grad():
    n_output = t_linear.native_forward(t_input)
    s_output = t_linear.forward_fp32_simulated(t_input)
    t_linear.post_process_weights()
    q_output = t_linear.forward(t_input)
print("s_diff", (n_output - s_output).abs().mean(), "q_diff", (n_output - q_output).abs().mean())
```

s_diff tensor(6.7442e-05, device='cuda:0') q_diff tensor(0.4397, device='cuda:0')

**Comment by LeiWang1999** [comment]
Thank you for your reports. Previously we made a pull request that make int1 represents the range [-1, 1], and int2 represents [-2, 1]. It seems this has introduced a bug in the bitnet integration. I'm currently working on other items, but I will address this issue in the coming days.

**Comment by LeiWang1999** [comment]
The pipeline and correctness tests passed on my main branch; could you please provide your BitBlas version and device details?

**Comment by igrekun** [comment]
BitBLAS version: bitblas==0.0.1.dev7
Devices tested: Nvidia T4, Nvidia A100.
Cuda version: 12.1
Ran the snippet above that computes mean difference. Should I test the main branch and build from source?

**Comment by LeiWang1999** [comment]
> Recent update might've broken something. Here are the results between simulated and bitblas matmul
> 
> ```
> t_input = torch.randn(1, 64).to("cuda")
> t_linear = utils_quant.BitLinear(64, 128, bias=False).to("cuda")
> with torch.no_grad():
>     n_output = t_linear.native_forward(t_input)
>     s_output = t_linear.forward_fp32_simulated(t_input)
>     t_linear.post_process_weights()
>     q_output = t_linear.forward(t_input)
> print("s_diff", (n_output - s_output).abs().mean(), "q_diff", (n_output - q_output).abs().mean())
> ```
> 
> s_diff tensor(6.7442e-05, device='cuda:0') q_diff tensor(0.4397, device='cuda:0')

would you mind provide the whole script to reproduce this result? :)

**Comment by igrekun** [comment]
Sure, put this anyhwere in integration/BitNet folder

```
import torch
from utils_quant import BitLinear

t_input = torch.randn(1, 64).to("cuda")
t_linear = BitLinear(64, 128, bias=False).to("cuda")
with torch.no_grad():
    n_output = t_linear.native_forward(t_input)
    s_output = t_linear.forward_fp32_simulated(t_input)
    t_linear.post_process_weights()
    q_output = t_linear.forward(t_input)
print("s_diff", (n_output - s_output).abs().mean(), "q_diff", (n_output - q_output).abs().mean())
```

**Comment by LeiWang1999** [comment]
Thanks @igrekun , pr https://github.com/microsoft/BitBLAS/pull/49 fixed this issue and you can apply it by installing 0.0.1.dev8

```
pip install bitblas==0.0.1.dev8
```

**Comment by igrekun** [comment]
Validated. Works perfectly on A100.
Gives different results on T4 (Compute capability 7.5) though this is not listed as supported so guess we can close this.

**Comment by LeiWang1999** [comment]
hi @igrekun what's the results on T4?

**Comment by LeiWang1999** [comment]
Thanks, closed :)

**Comment by TeaPoly** [comment]
```
GPU: NVIDIA RTX A4000
bitblas=0.0.1.dev8
```

Use the same example:
```python
import torch
from utils_quant import BitLinear

t_input = torch.randn(1, 64).to("cuda")
t_linear = BitLinear(64, 128, bias=False).to("cuda")
with torch.no_grad():
    n_output = t_linear.native_forward(t_input)
    s_output = t_linear.forward_fp32_simulated(t_input)
    t_linear.post_process_weights()
    q_output = t_linear.forward(t_input)
print("s_diff", (n_output - s_output).abs().mean(), "q_diff", (n_output - q_output).abs().mean())

```

Result is here:
```log
2024-06-17 14:28:03 [BitBLAS:INFO]: Database path /home/huanglk/.cache/bitblas does not exist, skipping loading operators from the database
2024-06-17 14:28:06 [BitBLAS:INFO]: Start fast tuning with dynamic range
BitBLAS Tuning done, appended operator to global_operator_cache.
s_diff tensor(6.4458e-05, device='cuda:0') q_diff tensor(nan, device='cuda:0')
```

**Comment by LeiWang1999** [comment]
@TeaPoly Hi, would you mind update the bitblas version to `0.0.1.dev12` and check the scripts again ?

**Comment by TeaPoly** [comment]
> @TeaPoly Hi, would you mind update the bitblas version to `0.0.1.dev12` and check the scripts again ?

`pip install bitblas==0.0.1.dev12` and `rm -r ~/.cache/bitblas`

The result is same.

**Comment by LeiWang1999** [comment]
@MekkCyber , Utilizing `bitblas.set_log_level("Debug")` can yield more detailed information.

```python
import torch
from utils_quant import BitLinear
import bitblas
bitblas.set_log_level("DEBUG")

t_input = torch.randn(1, 64).to("cuda")
t_linear = BitLinear(64, 128, bias=False).to("cuda")
with torch.no_grad():
    n_output = t_linear.native_forward(t_input)
    s_output = t_linear.forward_fp32_simulated(t_input)
    t_linear.post_process_weights()
    q_output = t_linear.forward(t_input)
print("s_diff", (n_output - s_output).abs().mean(), "q_diff", (n_output - q_output).abs().mean())
```

The scripts works fine on my A100 gpu and I've no idea about the architecture detail of NVIDIA A4000, does it support INT8 Tensor Core?

**Comment by TeaPoly** [comment]
> @MekkCyber , Utilizing `bitblas.set_log_level("Debug")` can yield more detailed information.
> 
> ```python
> import torch
> from utils_quant import BitLinear
> import bitblas
> bitblas.set_log_level("DEBUG")
> 
> t_input = torch.randn(1, 64).to("cuda")
> t_linear = BitLinear(64, 128, bias=False).to("cuda")
> with torch.no_grad():
>     n_output = t_linear.native_forward(t_input)
>     s_output = t_linear.forward_fp32_simulated(t_input)
>     t_linear.post_process_weights()
>     q_output = t_linear.forward(t_input)
> print("s_diff", (n_output - s_output).abs().mean(), "q_diff", (n_output - q_output).abs().mean())
> ```
> 
> The scripts works fine on my A100 gpu and I've no idea about the architecture detail of NVIDIA A4000, does it support INT8 Tensor Core?

It works fine for `NVIDIA GeForce RTX 3090`, 
```
q_diff tensor(5.8702e-05, device='cuda:0')
```
but result in `NVIDIA RTX A4000` and `NVIDIA RTX A6000` is wrong.
```
q_diff tensor(nan, device='cuda:0')
```
Debug logger in A4000 is here:
[a4000.log](https://github.com/user-attachments/files/15863606/a4000.log)

**Comment by LeiWang1999** [comment]
@TeaPoly , thank you for your report. It appears there are some bugs related to the target data structure because TVM lacks an A4000 Target Tag. Unfortunately, I don't possess an A4000 GPU to reproduce the issue. Perhaps running `export TVM_TARGET="cuda" python test.py` might resolve it.
